taxonomy/openai/client.py

```diff
@@ -1,25 +1,45 @@
-"""openai client"""
+"""CHAT_COMPLETION_API client"""
+import json
+import logging
 
-import openai
+import requests
+from requests.exceptions import ConnectTimeout
 
 from django.conf import settings
 
-openai.api_key = settings.OPENAI_API_KEY
+log = logging.getLogger(__name__)
 
 
 def chat_completion(prompt):
     """
-    Use chatGPT https://api.openai.com/v1/chat/completions endpoint to generate a response.
-
+    Pass message list to chat endpoint, as defined by the CHAT_COMPLETION_API setting.
     Arguments:
         prompt (str): chatGPT prompt
     """
-    response = openai.ChatCompletion.create(
-        model="gpt-3.5-turbo",
-        messages=[
-            {"role": "user", "content": prompt},
-        ]
-    )
+    completion_endpoint = getattr(settings, 'CHAT_COMPLETION_API', None)
+    completion_endpoint_key = getattr(settings, 'CHAT_COMPLETION_API_KEY', None)
+    if completion_endpoint and completion_endpoint_key:
+        headers = {'Content-Type': 'application/json', 'x-api-key': completion_endpoint_key}
+        connect_timeout = getattr(settings, 'CHAT_COMPLETION_API_CONNECT_TIMEOUT', 1)
+        read_timeout = getattr(settings, 'CHAT_COMPLETION_API_READ_TIMEOUT', 15)
+        body = {'message_list': [{'role': 'assistant', 'content': prompt},]}
+        try:
+            response = requests.post(
+                completion_endpoint,
+                headers=headers,
+                data=json.dumps(body),
+                timeout=(connect_timeout, read_timeout)
+            )
+            chat = response.json().get('content')
+        except (ConnectTimeout, ConnectionError) as e:
+            error_message = str(e)
+            connection_message = 'Failed to connect to chat completion API.'
+            log.error(
+                '%(connection_message)s %(error)s',
+                {'connection_message': connection_message, 'error': error_message}
+            )
+            chat = connection_message
+    else:
+        chat = 'Completion endpoint is not defined.'
 
-    content = response['choices'][0]['message']['content']
-    return content
+    return chat
```

Comment:
I think we should remove the `if..else` and `try..except` and allow to fail loudly. Setting default message will write the incorrect response in database. Please see the `chat_completion` usages in https://github.com/openedx/taxonomy-connector/blob/master/taxonomy/utils.py

Reply:
Done